### PR TITLE
[Fix] schedulefree

### PR DESCRIPTION
**Pull request by hatonosuke**
## Problem (Why?)

The schedule-free optimizer in pytorch-optimizer differs from the original implementations shown below.

https://github.com/facebookresearch/schedule_free

Below are the differences and questions.
Could you please review this? 

### weight decay

Since weight decay needs to be applied appropriately to both z and y, processing is required differently from general Adam.
I implemented processing for schedulefree.

### lr include beta2 bias correction

Original implementations of AdamW/RAdam schedulefree do not include beta2 bias correction in the weights used to calculate the mean of parameters. I removed bias_corretion2 from lr.

### use_palm

Original implementations does not have use_palm.
I don't know what kind of behavior use_palm is trying to implement because it's not documented.
Therefore, I predicted beta2  to be a scheduling method of (1 - step**-0.8) in the following paper.

https://arxiv.org/abs/2204.02311

(1 - step**-0.8) is 0 when step is 1, I thought it could be implemented by simply removing bias correction and setting beta2 to (1-step**-0.8).

However, as in the PaLM implementation below, it is possible to always change betaes without any problems.
I think that if this implementation run while changing beta2, the behavior of this PR will be slightly different.

https://github.com/conceptofmind/PaLM/blob/c95d8c42fdb957eb8ec506954f74dee9aa184b5c/palm/stable_adamw.py#L71

### RAdam and use_plam

When use_palm is True, scheduled beta2 must also be used to calculate the coefficients of RAdam.
However, I do not have the ability to derive a reasonable RAdam calculation formula when scheduling beta2.

### ScheduleFreeRAdam when degenerated_to_sgd=False

I change that ScheduleFreeRAdam set lr to 0 when n_sma < 4 and degenerated_to_sgd is False.
And I add tests with degenerated_to_sgd=False.

## Other changes (bug fixes, small refactors)

I applied bias_correction2 before ams_bound.
I believe this behavior is correct in AMSGrad, but elsewhere in this repository bias_correction2 is applied after ams_bound.

Is this the intended behavior?

And since weight_decouple has no meaning in SGD, I removed weight_decouple ScheduleFreeSGD.

### CODE REVIEW

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 1]
```diff
@@ -271,9 +262,8 @@ def step(self, closure: CLOSURE = None) -> LOSS:
 
             if group['use_palm']:
                 beta2: float = 1.0 - group['step'] ** -0.8
-                debias: float = (1.0 - beta2) / (1.0 - beta2 ** group['step'])
-            else:
-                debias: float = beta2
+                # unnecessary bias correction when PaLM beta2 scheduling
+                bias_correction2 = 1.0
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/schedulefree.py:266]: yeap. original implementation does not have `PaLM` variant, which is from [here](https://github.com/ClashLuke/schedule_free/blob/main/schedulefree/palm_sfadamw.py).

I think `debias = beta2` is correct when `use_palm` is False (== `original sfadamw`), 
and need to fix to `debias: float = 1.0 - (1.0 - beta2) / (1.0 - beta2 ** group['step'])` in case of `use_palm` to align with the reference and `bias_correction2 = 1.0` should also be moved to under `if group['use_palm']:` condition!

**Review by kozistr — COMMENTED** [review]
first of all, thanks for your contributions! i commented on your reviews, and please feel free to leave comments or ask questions :)

I haven't answered everything yet, but I'll review them maybe later today.

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 2]
```diff
@@ -428,6 +414,12 @@ def step(self, closure: CLOSURE = None) -> LOSS:
                 n_sma_threshold=4,
                 degenerated_to_sgd=group['degenerated_to_sgd'],
             )
+            if n_sma > 4:
+                # cancel bias correction2
+                lr = lr / bias_correction2_sq
+            elif lr < 0.:
+                # n_sma < 4.0 and degenerated_to_sgd is False
+                lr = 0.0
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/schedulefree.py:422]: it seems like your modification is different from the original ScheduleFreeRAdam optimizer [here](https://github.com/facebookresearch/schedule_free/blob/main/schedulefree/radam_schedulefree.py#L163). could you elaborate more on this?

as far as i know, usually in [`RAdam` optimizer](https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L156), parameter is not gonna updated in case of `lr` is negative (like your comment, n_sma is under the threshold and `degenerated_to_sgd` is `False`).

(of course, i mistakenly miss to handle that case in the current code), i think it'd be more efficient to skip the parameter update instead of setting the lr to 0.0!

**Comment by kozistr** [comment]
> I applied bias_correction2 before ams_bound. I believe this behavior is correct in AMSGrad, but elsewhere in this repository bias_correction2 is applied after ams_bound.

yes. as you said, bias_correction2 is applied before ams_bound. however, in my implementation, i thought that it brought a little bit of efficient computation, dividing `de_nom` with sqrt of `bias_correction2`, which is almost the same result i guess.

however, on second thought, there seems no advantage in doing that, so gonna refactor that later.

thanks for brining this up :)

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 3]
```diff
@@ -30,7 +29,6 @@ def __init__(
         lr: float = 1.0,
         momentum: float = 0.9,
         weight_decay: float = 0.0,
-        weight_decouple: bool = True,
```
[the remark this answers is not in the document]
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:33]: I understood about weight_decouple is for API consistency.
When weight_decouple is False, the logic is identical to the original implementation. But when weight_decouple is True, the logic is not identical. Details will be explained later.

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 4]
```diff
@@ -259,9 +250,9 @@ def step(self, closure: CLOSURE = None) -> LOSS:
 
             beta1, beta2 = group['betas']
 
-            bias_correction2_sq: float = math.sqrt(1.0 - beta2 ** group['step'])
+            bias_correction2: float = 1.0 - beta2 ** group['step']
 
-            lr: float = group['lr'] * schedule * bias_correction2_sq
+            lr: float = group['lr'] * schedule
```
[the remark this answers is not in the document]
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:255]: I didn't know that original was implemented like that before. 
Thank you!

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 5]
```diff
@@ -289,31 +279,27 @@ def step(self, closure: CLOSURE = None) -> LOSS:
...
                     state['exp_avg_sq'] = torch.zeros_like(p)
 
-                self.apply_weight_decay(
-                    p=p,
-                    grad=grad,
-                    lr=lr,
-                    weight_decay=group['weight_decay'],
-                    weight_decouple=group['weight_decouple'],
-                    fixed_decay=group['fixed_decay'],
-                )
```
[the remark this answers is not in the document]
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:299]: In schedulefree, "weight decay" is the process that not "y" but "z" is decayed by "y"("p" in code).
But BaseOptimizer.apply_weight_decay() decays "y"("p" in code) when decouple_weight is True.

In addition, $y_t$ is calculated from $z_{t-1}$, $y_{t-1}$ and $(z_{t-1} - z_{t})$("grad" in code)$.
Therefore, "z" should not change before update "y".

Therefore, "weight_decay" should first be applied to $(z_{t-1}-z_{t})$ ("grad" in code) and indirectly to z and y through $z_{t-1}-z_{t}$.

File: pytorch_optimizer/optimizer/schedulefree.py [hunk 6]
```diff
@@ -289,31 +279,27 @@ def step(self, closure: CLOSURE = None) -> LOSS:
...
+                grad.mul_(lr)
+                if group['weight_decouple']:
+                    grad.add_(p, alpha=group['weight_decay'] * (1.0 if group['fixed_decay'] else lr))
 
                 p.lerp_(z, weight=checkpoint)
-                p.add_(grad, alpha=lr * (beta1 * (1.0 - checkpoint) - 1))
+                p.add_(grad, alpha=beta1 * (1.0 - checkpoint) - 1)
 
-                z.sub_(grad, alpha=lr)
+                z.sub_(grad)
```
[the remark this answers is not in the document]
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:302]: I think your idea is cool. However, the final value of p is "x" instead of "y".
Thus the following is required last:
```python
p.leap_(end=z, weight=1.0 - beta1)   # x to y
```
Although it sacrifices clarity, it is possible to combine to one lerp:
```python
p.leap_(end=z, weight=(1.0 - beta1) * (1 + checkpoint))
```
However, even with this code, weight decay shroud decay "z" by "y".
So  before convert  "y" to "x", "weight decay" need to be applied to "grad" and then indirectly to "z".

File: pytorch_optimizer/optimizer/schedulefree.py (hunk 1 above)
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:266]: I checked again and undarstand that $\hat{v} _t$ can be calculated from $\hat{v} _{t - 1}$ by using the original implementation of debias.

Thank you!

File: pytorch_optimizer/optimizer/schedulefree.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply pytorch_optimizer/optimizer/schedulefree.py:299]: yeap. When `decouple_weight` is `True` (which is mistakenly set), it doesn't perform as intended. but, when it is `False`, it becomes identical.

well, maybe, it'd be better to remove the `decoupled_weight` parameter not to change its value.

File: pytorch_optimizer/optimizer/schedulefree.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply pytorch_optimizer/optimizer/schedulefree.py:302]: I think that setting `weight decouple` to `False` and moving `apply_weight_decay` after `de_nom` solves this issue too.

```
# pseudo-code
# zt+1 = zt − γtgt/(√vt + ϵ) − γtλyt -> (z.sub_(grad, lr))
# xt+1 = (1 − ct+1) xt + ct+1zt+1 -> (p.lerp(z, checkpoint)

grad = ... # normalized (coupled-style) weight-decayed gradient

# https://github.com/facebookresearch/schedule_free/blob/main/schedulefree/adamw_schedulefree.py#L204 # same as current implementation here
p.lerp_(z, weight=checkpoint)
p.add_(grad, alpha=lr * (beta1 * (1.0 - checkpoint) - 1))
z.sub_(grad, alpha=lr)

or also can be like below

# https://github.com/facebookresearch/schedule_free/blob/main/schedulefree/adamw_schedulefree_closure.py # pseudo-code friendly
p.lerp_(end=z, weight=1.0 - 1.0 / beta1)
z.sub_(grad, alpha=lr)
p.lerp_(end=z, weight=checkpoint)
```

File: pytorch_optimizer/optimizer/schedulefree.py (hunk 2 above)
Comment (kozistr) [reply pytorch_optimizer/optimizer/schedulefree.py:422]: i get your point! well, I agree that even if the parameters aren't updated, setting lr to 0.0 or sth else seems great.

File: pytorch_optimizer/optimizer/schedulefree.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hatonosuke) [reply pytorch_optimizer/optimizer/schedulefree.py:308]: When I tried to add weigth_decay to grad for "fixed_decay", the implementation was like that.

**Comment by kozistr** [comment]
@hatonosuke I checked all of your work and reviews, and seems like there's kinda a lot to change and refactor. So, I'm thinking of closing this PR and opening another branch, then working there based on this PR. (of course, the credit all goes to you)

Here are things to work on:
1. fix weight decay issue (maybe I'm gonna remove the decouple wd option)
2. update the sf optimizers with the latest one
3. fix when step size is negative in sfradam
4. refactor palm part
5. refactor amsbound part

thanks for your contributions! and if you have more questions or reviews, please feel free to leave here :)

**Comment by hatonosuke** [comment]
I got it.
Thank you for your responses!
